### ISSUE TITLE
Share extensions will *only* show up if they explicitly support *all* of the provided activity items

### ISSUE THREAD

**Issue by irace** [tumblr/ios-extension-issues#5]
* [Radar #18342403: NSExtensionActivationRules should only need to match a single activity item for a share extension to be displayed](http://openradar.appspot.com/radar?id=5616559737274368)
* [Radar #18150467: Documentation for custom NSExtensionItemActivation rules is very vague](http://openradar.appspot.com/radar?id=5803657102622720)

This is a doozy. It’s the most important issue we’ve found, and one that probably deserves a blog post of its own.

Here’s how applications pass data to share extensions: 

* An application configures a `UIActivityViewController` with an array of “activity items”
* The activity controller displays the system activities and share extensions that can operate on the types of items provided

Here’s how we think this should work, using the Tumblr app as an example:

* The user long-presses on a photo
* We put the image data, the posts’s URL, and maybe a text summary of the post, all in the activity items array
* We’d expect share extensions that support either image data *or* URLs *or* text to all show up in the activity controller

What *actually* happens is that only share extensions that explicitly support images *and* URLs *and* text will show up. 

This is a problem, because the simplest way to specify what your extension supports – and by far the best documented – is by adding `NSExtensionActivationRule` keys like:

    `NSExtensionActivationSupportsText` : `YES`

This looks like it would mean “show my extension as long as *any* of the activity items are text,” but it really means “show my extension as long as there is only one activity item, and it is text.”

[Federico Viticci](http://twitter.com/viticci), who at this point has likely used more third-party share extensions than anyone else on the planet, verifies that this is in fact a legitimate problem:

<blockquote class="twitter-tweet" lang="en"><p><a href="https://twitter.com/irace">@irace</a> Yup. Been talking to devs to handle exceptions when possible, but I&#39;m getting a lot of failures in several apps.</p>&mdash; Federico Viticci (@viticci) <a href="https://twitter.com/viticci/status/507992803872616448">September 5, 2014</a></blockquote>
<blockquote class="twitter-tweet" lang="en"><p><a href="https://twitter.com/irace">@irace</a> Yep. And, that the input passed by an app doesn&#39;t match what another app&#39;s extension expects and you get all sorts of weird stuff.</p>&mdash; Federico Viticci (@viticci) <a href="https://twitter.com/viticci/status/507996744865837056">September 5, 2014</a></blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>

### Workaround

This negatively affects both app and extension developers. It means that:

* App developers should only configure their activity controllers with a single activity item. There are a couple of problems with this. 
    1) [It’s doable](https://gist.github.com/irace/535c9aa3314ee41fb902), but a pain if, like in Tumblr.app, you want system activities like copying and saving to the Camera Roll to support multiple different types of data. 
    2) It’s a huge shame to only export one type of data and limit the number of sharing options that your users will be able to perform.

* Extension developers should use the more complex (and unfortunately, not very thoroughly documented) [predicate syntax](https://developer.apple.com/library/prerelease/ios/documentation/General/Conceptual/ExtensibilityPG/ExtensionScenarios.html#//apple_ref/doc/uid/TP40014214-CH21-SW8) to specifically specify an OR relationship. This would look something like:

<code>SUBQUERY(extensionItems, $extensionItem, SUBQUERY($extensionItem.attachments, $attachment, ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.image").@count <= 10).@count >= 1 OR 
SUBQUERY(extensionItems, $extensionItem, SUBQUERY($extensionItem.attachments, $attachment, ANY $attachment.registeredTypeIdentifiers UTI-CONFORMS-TO "public.text").@count >= 1).@count >= 1</code>

**Comment by irace** [comment]
This is fixed in iOS 9, as per the [App Extension Best Practices](https://developer.apple.com/videos/wwdc/2015/?id=224) session :tada:.

To opt your extension into OR instead of AND behavior, simply add the following key/value pair to your Info.plist:

```
NSExtensionActivationDictionaryVersion: 2
```

**Comment by jasewells** [comment]
> The user long-presses on a photo
> We put the image data, the posts’s URL, and maybe a text summary of the post, all in the activity items array

Anyone know if the above is the actual behavior of the Tumblr app? I have an action extension that accepts an image, but cannot get it to show up when you long-press a photo. I’ve got NSExtensionActivationSupportsImageWithMaxCount = 1 and NSExtensionActivationDictionaryVersion = 2 in my Info.plist and my extension does appear when doing a similar share from Photos, Twitter, Flickr, etc. But Tumblr only seems to provide a public.url attachment type and nothing matching public.image. (And the link it provides is to the post, not a specific image.) So I’m just wondering if the above idea of providing an image, link, and text together is really being done or if that was just an example/idea and not actually how Tumblr provides activity items.

Incidentally, the standard system “Save Photo” and “Copy” actions _do_ appear from Tumblr, so somehow they know there is image data?

**Comment by paulrehkugler** [comment]
You're right; that's not the behavior of the Tumblr app.

In order to avoid having photo posts be unable to be shared by share extensions that do not explicitly support images, we opted for what we assumed to be the lowest common denominator for all share extensions: URLs.

The system copy activity simply acts on the URL of the photo post; if you choose Copy and then paste in a text area, you'll see it's a URL.

The "Save Photo" activity is trickier. Because the system save image activity will not show up based on a URL, we wrote our own custom save image activity that we use when long pressing photos.
